index/ask/runner/runner.go

```diff
@@ -0,0 +1,299 @@
+	index := ask.Index{
+		LastUpdated:        ai.index.LastUpdated,
+		StorageMedianPrice: ai.index.StorageMedianPrice,
+		Storage:            make(map[string]ask.StorageAsk, len(ai.index.Storage)),
+	}
+	for addr, v := range ai.index.Storage {
+		index.Storage[addr] = v
+	}
+	return index
+}
+
+// Query executes a query to retrieve active Asks
+func (ai *Runner) Query(q Query) ([]ask.StorageAsk, error) {
+	ai.lock.Lock()
+	defer ai.lock.Unlock()
+	var res []ask.StorageAsk
+	offset := q.Offset
+	for _, sa := range ai.orderedAsks {
+		if q.MaxPrice != 0 && sa.Price > q.MaxPrice {
+			break
+		}
+		if q.PieceSize != 0 && sa.MinPieceSize > q.PieceSize {
+			continue
+		}
+		if offset > 0 {
+			offset--
+			continue
+		}
+		res = append(res, *sa)
+		if q.Limit != 0 && len(res) == q.Limit {
+			break
+		}
+	}
+	return res, nil
+}
+
+// Listen returns a new channel signaler that notifies when the index gets
+// updated.
+func (ai *Runner) Listen() <-chan struct{} {
+	return ai.signaler.Listen()
+}
+
+// Unregister unregisters a channel signaler from the signaler hub
+func (ai *Runner) Unregister(c chan struct{}) {
+	ai.signaler.Unregister(c)
+}
+
+// Close closes the AskIndex
+func (ai *Runner) Close() error {
+	log.Info("Closing")
+	ai.clsLock.Lock()
+	defer ai.clsLock.Unlock()
+	if ai.closed {
+		return nil
+	}
+	ai.cancel()
+	<-ai.finished
+	ai.signaler.Close()
+	ai.closed = true
+	return nil
+}
+
+// start is a long running job that updates asks information in the market
+func (ai *Runner) start() {
+	defer close(ai.finished)
+	if err := ai.update(); err != nil {
+		log.Errorf("error when updating miners asks: %s", err)
+	}
+	for {
+		select {
+		case <-ai.ctx.Done():
+			log.Info("graceful shutdown of ask index background job")
+			return
+		case <-time.After(qaRefreshInterval):
+			if err := ai.update(); err != nil {
+				log.Errorf("error when updating miners asks: %s", err)
+			}
+		}
+	}
+}
+
+// update triggers a full-scan generates and saves a new fresh index and builds
+// views for better querying.
+func (ai *Runner) update() error {
+	log.Info("updating ask index...")
+	defer log.Info("ask index updated")
+	startTime := time.Now()
+	newIndex, cache, err := generateIndex(ai.ctx, ai.api)
+	if err != nil {
+		return fmt.Errorf("generating index: %s", err)
+	}
+
+	ai.lock.Lock()
+	ai.index = newIndex
+	ai.orderedAsks = cache
+	ai.lock.Unlock()
+
+	stats.Record(context.Background(), metrics.MFullRefreshDuration.M(time.Since(startTime).Milliseconds()))
+	ai.signaler.Signal()
+
+	return nil
+}
+
+// generateIndex returns a fresh index
+func generateIndex(ctx context.Context, api *apistruct.FullNodeStruct) (ask.Index, []*ask.StorageAsk, error) {
+	addrs, err := api.StateListMiners(ctx, types.EmptyTSK)
+	if err != nil {
+		return ask.Index{}, nil, err
+	}
+
+	rateLim := make(chan struct{}, qaRatelim)
+	var lock sync.Mutex
+	newAsks := make(map[string]ask.StorageAsk)
+	for i, addr := range addrs {
+		if ctx.Err() != nil {
+			break
+		}
```

Comment:
This is a new thing; no that much interesting.
Sometimes if you ctrl+c, you'd have to spend quite a bit of time waiting for a current index rebuild to fuinish.
Adding this, the loop is cut out so it can return faster to cleanly close up powd.